packages/ogimage-gen/src/OgImageMiddleware.ts

```diff
@@ -0,0 +1,344 @@
+
+  // Initialized in invoke() 👇
+  imageProps?: {
+    width: number
+    height: number
+    quality: number
+  }
+
+  constructor(options: MwOptions) {
+    this.options = options
+
+    this.App = options.App
+    this.Document = options.Document
+  }
+
+  async invoke(
+    req: MiddlewareRequest,
+    mwResponse: MiddlewareResponse,
+    invokeOptions: MiddlewareInvokeOptions,
+  ) {
+    const url = new URL(req.url)
+    const { pathname } = url
+
+    let currentRoute: RWRouteManifestItem | undefined = undefined
+    let parsedParams: {
+      params?: Record<string, unknown>
+    } = {}
+
+    // Skip processing if not a file request
+    if (!pathname.includes('.')) {
+      return mwResponse
+    }
+
+    // Remove the extension for the match
+    const [routePathname, extension] = pathname.split('.')
+
+    ;({ currentRoute, parsedParams } = await this.matchRoute(
+      // This is a special case for the index route
+      // because they can't go to mywebsite.com/.png -> mywebsite.com/index.png instead
+      routePathname === '/index' ? '/' : routePathname,
+      parsedParams,
+    ))
+
+    // If no match with the router, or not a supported extension bail
+    if (!currentRoute || !supportedExtensions.includes(extension)) {
+      console.log(
+        'OGMiddleware: No match with the Routes, or not a supported extension',
+      )
+      return mwResponse
+    }
+
+    // Combine search params and params from route pattern
+    // /user/{id:Int} => /user/1?background=red => { id: 1, background: 'red'}
+    const mergedParams = {
+      ...Object.fromEntries(url.searchParams.entries()),
+      ...(parsedParams.params || {}),
+    }
+
+    this.imageProps = {
+      width: mergedParams.width
+        ? parseInt(mergedParams.width as string)
+        : OGIMAGE_DEFAULTS.width,
+      height: mergedParams.height
+        ? parseInt(mergedParams.height as string)
+        : OGIMAGE_DEFAULTS.height,
+      quality: mergedParams.quality
+        ? parseInt(mergedParams.quality as string)
+        : OGIMAGE_DEFAULTS.quality,
+    }
+
+    const debug = !!mergedParams.debug
+
+    const screenshotOptions: ScreenshotOptions = {
+      viewport: {
+        width: this.imageProps.width,
+        height: this.imageProps.height,
+      },
+      format: {
+        png: { type: 'png' },
+        jpg: {
+          type: 'jpeg',
+          quality: this.imageProps.quality,
+        },
+      },
+    }
+
+    // @TODO (ROB!)
+    // I think it doesn't work with jsx paths in my project IN DEV at the moment
+    // Try renaming AboutPage.png.tsx -> AboutPage.png.jsx
+    const ogImgFilePath = this.getOgComponentPath(currentRoute, extension)
+
+    const { data, Component } = await this.importComponent(
+      ogImgFilePath,
+      invokeOptions,
+    )
+
+    let dataOut
+    if (data && typeof data === 'function') {
+      dataOut = await data(mergedParams)
+    }
+
+    // @TODO (ROB): We could pass in playwright as an option to the middleware.
```

Comment:
So in entry.server you'd pass App, Document and playwright to new OgImageMiddleware()? Hmm...does it effect bundle size? Or does it just feel better that you're passing the dependency instead of having this class require import it itself?

If the rendering code was somehow generic enough that it could use any renderer, and playwright is just one possibility, I think that would be awesome: you pass renderer at instantiation time and it just uses that to render the image, but there isn't currently any other option, is there? Didn't our testing show that only Playwright was able to render everything correctly (since it was a real web browser and parsing CSS the same way, unlike that Vercel package which supported a subset of CSS)?

Reply:
> So in entry.server you'd pass App, Document and playwright to new OgImageMiddleware()?

Yeah exactly this. It does not affect bundle size, because this is all on the server. It's mainly a way of making it clear that the user will need to install playwright, then pass it to us. So both `yarn install playwright` and `npx playwright install` isn't really a middleware setup step then, more something they need to do on their project. Maybe I'm just overcomplicating things though. 

> If the rendering code was somehow generic enough
haha lets not go there.... but yes a possibility in the future.

Reply:
See readme for playwright on how to install. https://www.npmjs.com/package/playwright

I suppose we have the same in OgMiddleware? To use you need to run this stuff?

Reply:
Yeah that makes sense. After running `yarn rw setup middleware ogimage` we can have a simplified version of those install instructions in the last step.